api/provider/bitbucket.go

```diff
@@ -0,0 +1,63 @@
+package provider
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/netlify/gotrue/conf"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/bitbucket"
+)
+
+// Bitbucket
+
+type bitbucketProvider struct {
+	*oauth2.Config
+}
+
+func NewBitbucketProvider(ext conf.ExternalConfiguration) Provider {
+	return &bitbucketProvider{
+		&oauth2.Config{
+			ClientID:     ext.Key,
+			ClientSecret: ext.Secret,
+			Endpoint:     bitbucket.Endpoint,
+		},
+	}
+}
+
+func (g bitbucketProvider) GetOAuthToken(ctx context.Context, code string) (*oauth2.Token, error) {
+	return g.Exchange(ctx, code)
+}
+
+func (g bitbucketProvider) GetUserEmail(ctx context.Context, tok *oauth2.Token) (string, error) {
```

Comment:
Yeah, I will definitely refactor to use a struct to avoid the inference stuff.

The issue here is that Bitbucket uses a url like "https://bitbucket.org/users/<USERNAME>/emails" - So the first request is just to get the logged in user's username in order to be able to make the request to Bitbucket for their email, and the second request to get the email addresses follows the same pattern as Github, which is why I was able to abstract that away into getUserEmail.

I can definitely work on cleaning this up a little bit more though.